src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -88,6 +89,23 @@ cagg_watermark_get(Hypertable *mat_ht)
 	ScanIterator iterator =
 		ts_scan_iterator_create(CONTINUOUS_AGGS_WATERMARK, AccessShareLock, CurrentMemoryContext);
 
+	/* Planner must have a snapshot in case it calls user-defined functions (see pg_plan_query()) */
+	Assert(ActiveSnapshotSet());
+
+	/*
+	 * The watermark of a CAGG has to be fetched by using the active snapshot.
+	 *
+	 * The watermark of a CAGG can be read by multiple parallel workers (i.e., in ChunkAppend). All
+	 * of the workers have to see the same watermark in order to act consistently (e.g., when
+	 * excluding chunks). If no snapshot is specified, there is a chance that one worker may see a
+	 * different watermark than another, if the CAGG was updated in between their reads (e.g., read
+	 * of worker a, CAGG update, read of worker B).
+	 *
+	 * PostgreSQL synchronizes the snapshots between the parallel worker. Using the same snapshot
+	 * ensures that all parallel workers have the same view of the stored watermark.
+	 */
+	iterator.ctx.snapshot = GetActiveSnapshot();
```

Comment:
I think this logic should be moved inside `scanner` and the less likely used concurret stuff should be the exceptional case - so that we are clear of troubles unless we want to face them

Reply:
Yes, a normal MVCC snapshot should be the default behavior. I created a follow-up ticket for this.